File: Java/src/GraphMST.java
import java.util.*;

public class GraphMST {

    public static class Edge implements Comparable<Edge> {
        public int u,v,w;
        public Edge(int u,int v,int w) {this.u=u;this.v=v;this.w=w;}
        public int compareTo(Edge e){return Integer.compare(this.w, e.w);}
    }

    public static List<Edge> kruskal(int n, List<Edge> edges){
        Collections.sort(edges);
        UnionFind uf = new UnionFind(n);
        List<Edge> mst = new ArrayList<>();
        for (Edge e : edges) {
            if (uf.union(e.u, e.v)) {
                mst.add(e);
                if (mst.size()==n-1) break;
            }
        }
        return mst;
    }
}

File: Python/src/graph_mst.py
# Kruskal and Prim
from typing import List, Tuple
import heapq
from .graph_basic import UnionFind

def kruskal(n: int, edges: List[Tuple[int,int,int]]) -> List[Tuple[int,int,int]]:
    # edges: list of (u,v,w)
    edges_sorted = sorted(edges, key=lambda x: x[2])
    uf = UnionFind(n)
    mst = []
    for u,v,w in edges_sorted:
        if uf.union(u,v):
            mst.append((u,v,w))
        if len(mst) == n-1:
            break
    return mst

def prim(n: int, adj: List[List[Tuple[int,int]]], start=0):
    # adj[u] = list of (v, weight)
    visited = [False]*n
    pq = [(0, start, -1)]  # (w, u, parent)
    mst = []
    while pq:
        w,u,parent = heapq.heappop(pq)
        if visited[u]: continue
        visited[u] = True
        if parent != -1:
            mst.append((parent, u, w))
        for v,wt in adj[u]:
            if not visited[v]:
                heapq.heappush(pq, (wt, v, u))
    return mst
